### ISSUE TITLE
Selectors

### ISSUE THREAD

**Issue by zeebo** [go-goods/tmpl#4]
Currently the idea for selectors is to have push and pop operations on identifiers, so for example, some parsings

```
.foo.bar$ -> push foo push bar pop
$$foo -> pop pop foo
```

Now I can normalize these operations such that it starts with a number of pops, then a chain of push ident. So

```
.foo.bar$ -> push foo
$$foo -> pop pop foo
```

That's all fine. the question is about sub contexts in blocks, with, and range statements and how they interact with pop. Assuming the context is a dict like

```
{
    'foo': 'foo',
    'bar': {
        'foo': 'baz',
        'bam': {
            'zap': 'zap'
        }
    }
}
```

what is the output of something like

```
{% with .bar.bam %}{% .zap %}{% end with %} -> 'zap'
{% with .bar.bam %}{% $foo %}{% end with %} -> 'baz' or 'foo' ?
```

Basically, is it popping off the last context element (.bar.bam) or the last push element (.bam)?
same for ranges and blocks.

**Comment by zeebo** [comment]
Another thought. Should we make pops require a push after? So $$foo would be invalid, but $$.foo would not.

I feel like this might make the issue more clear. doing {% with .bar.bam %} would be like the rooted path /bar/bam, and {% $.foo %} would be like /bar/bam/../foo, or /bar/foo or baz.

**Comment by zeebo** [comment]
Decided to make pushes come after pops, so that push is really more of an access command. with and block will push contexts on to the stack, so that for the second example, it would output "baz". Ranges will use the new variable accessors.

Would love to hear comments, reopen with any concerns.
